rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/geo/TbGpsMultiGeofencingActionNode.java

```diff
@@ -0,0 +1,172 @@
+        return null;
+    }
+
+    protected List<EntityId> findMatchedZones(TbContext tbContext, TbMsg tbMsg) throws TbNodeException, ExecutionException, InterruptedException {
+        JsonObject msgDataObj = getJsonObject(tbMsg);
+        double latitude = getValueFromMessageByName(tbMsg, msgDataObj, config.getLatitudeKeyName());
+        double longitude = getValueFromMessageByName(tbMsg, msgDataObj, config.getLongitudeKeyName());
+
+        if (config.getEntityRelationsQuery() == null) {
+            return Collections.emptyList();
+        }
+
+        EntityRelationsQuery entityRelationsQuery = config.getEntityRelationsQuery();
+
+        if (config.getEntityRelationsQuery().getParameters().getRootId() == null || config.getEntityRelationsQuery().getParameters().getRootType() == null) {
+            entityRelationsQuery = copyEntityRelationsQuery(tbMsg.getOriginator(), config.getEntityRelationsQuery());
+        }
+
+        List<EntityId> relatedZones = getInFuture(tbContext.getRelationService().findByQuery(tbContext.getTenantId(), entityRelationsQuery)).stream().map(this::getEntityIdFromEntityRelation).toList();
+
+        Map<EntityId, Perimeter> entityIdPerimeterMap = new HashMap<>();
+
+        List<EntityId> matchedGeofenceIds = new ArrayList<>();
+
+        for (EntityId relatedGeofenceId : relatedZones) {
+            Optional<AttributeKvEntry> attributeKvEntry = tbContext.getAttributesService().find(tbContext.getTenantId(), relatedGeofenceId, AttributeScope.SERVER_SCOPE, config.getPerimeterAttributeKey()).get();
+            attributeKvEntry.map(this::mapToPerimeter).ifPresent(p -> entityIdPerimeterMap.put(relatedGeofenceId, p));
+        }
+
+        for (EntityId entityId : entityIdPerimeterMap.keySet()) {
+            Perimeter perimeter = entityIdPerimeterMap.get(entityId);
+            boolean matches = checkMatches(perimeter, latitude, longitude);
+            if (matches) {
+                matchedGeofenceIds.add(entityId);
+            }
+        }
+
+        return matchedGeofenceIds;
+    }
+
+    private EntityRelationsQuery copyEntityRelationsQuery(EntityId originatorId, EntityRelationsQuery entityRelationsQuery) {
+        EntityRelationsQuery copy = new EntityRelationsQuery();
+        copy.setParameters(entityRelationsQuery.getParameters());
+        copy.getParameters().setRootId(originatorId.getId());
+        copy.getParameters().setRootType(originatorId.getEntityType());
+        copy.setFilters(entityRelationsQuery.getFilters());
+        return copy;
+    }
+
+    private EntityId getEntityIdFromEntityRelation(EntityRelation entityRelation) {
+        return EntitySearchDirection.TO.equals(config.getRelationField()) ? entityRelation.getTo() : entityRelation.getFrom();
+    }
+
+    private <T> T getInFuture(ListenableFuture<T> future) {
+        try {
+            return future.get();
+        } catch (InterruptedException | ExecutionException e) {
+            throw new RuntimeException(e);
+        }
+    }
```

Comment:
Blocking operations is not allowed in the rule nodes. 

> We can't do blocking operations .get() in the rule nodes since these block the rule node dispatcher threads from processing other nodes. You should use callbacks to handle future results. Please take a look at how we handle futures in other rule nodes.